### DISCUSSION TITLE
Backfill: report assigned tickets in acknowledge response?

### DISCUSSION THREAD

**Discussion by scosgrave** [googleforgames/open-match#1360] [Ideas]
The current way that game servers can learn what players will be back-filled into them is through the Director sending match information to the game server or to a service that manages which players are in a game. This tells the game server which players have been matched to the backfill, but there isn't a very strong guarantee that the players will actually arrive because an AcknowledgeBackfillRequest still needs to be sent by the server to Open Match, and those same tickets from the match the Director sent need to still be on the Backfill ticket when the acknowledge happens. If the game server takes too long to send the acknowledge, the players associated with the backfill could have been released from the backfill. Some players may have disconnected from Open Match as well prior to the acknowledge arriving, and their ticket is no longer associated with the backfill.

There are a few ways to strengthen the guarantee that the tickets on the backfill will get an assignment and those game clients will arrive at the game server:

1. If an assignment has happened on the Backfill in the past, that information is saved, and any future matches put on the Backfill will automatically get that assignment info. This would speed up the assignment of tickets on Backfills, but it takes away the game server's ability to control the rate of players connecting, and its ability to not let players in if the state of the world on the game server has changed, and it no longer wants the type of players it had previously asked for.
2. We can update the AcknowledgeBackfill API's response to include the ticket IDs that got the assignment. This list of ticket IDs will represent clients that are more likely to connect than the ones that the Director sent when it reported the match.  
3. The Frontend can keep track of player match requests, and assignments, and call your game session management service, or player state management service, when it sees an assignment happen. One problem with this approach is that it is not specific to Backfills. Also, it means the Frontend is taking on the same role the Director has in reporting the players that are joining a match, but with less information available to it than the Director.

Out of the options above, I think the 2nd option sounds the best so far, but I'd like to hear what others in the community think. Are there better options? Is option 1 or 3 better in your opinion? Do we really need to increase the accuracy of Open Match reporting which players will arrive at the game server?

**Comment by HazzaBui** [comment]
I'm actually facing this exact problem/situation right now, and I would definitely agree that option 2 feels like the best approach.  It is also the approach I'm trying to emulate currently, through adding the tickets to the backfill as encoded data in SearchFields.StringArgs, and reading it out on the gameserver - my current approach feels a bit brittle though, and I'd love to replace it with an OpenMatch built-in function

As a related aside, reading the OpenMatch website docs, it seems like option 2 you described above is actually how the functionality of the backfill feature is described? https://open-match.dev/site/docs/guides/backfill/#acknowledging-backfills  Specifically, in relation to the AcknowledgeBackfill call, the docs say `The Game Server receives the current Backfill status as well as a list of new tickets that were just acknowledged by the request as a response`

Edit: It looks like option 2 has already been implemented in the main branch, in #1382
